crates/ide-assists/src/handlers/bool_to_enum.rs

```diff
@@ -0,0 +1,728 @@
+use ide_db::{
+    assists::{AssistId, AssistKind},
+    defs::Definition,
+    search::{FileReference, SearchScope, UsageSearchResult},
+    source_change::SourceChangeBuilder,
+};
+use syntax::{
+    ast::{
+        self,
+        edit::IndentLevel,
+        edit_in_place::{AttrsOwnerEdit, Indent},
+        make, HasName,
+    },
+    ted, AstNode, NodeOrToken, SyntaxNode, T,
+};
+
+use crate::assist_context::{AssistContext, Assists};
+
+// Assist: bool_to_enum
+//
+// This converts boolean local variables, fields, constants, and statics into a new
+// enum with two variants `Bool::True` and `Bool::False`, as well as replacing
+// all assignments with the variants and replacing all usages with `== Bool::True` or
+// `== Bool::False`.
+//
+// ```
+// fn main() {
+//     let $0bool = true;
+//
+//     if bool {
+//         println!("foo");
+//     }
+// }
+// ```
+// ->
+// ```
+// fn main() {
+//     #[derive(PartialEq, Eq)]
+//     enum $0Bool { True, False }
+//
+//     let bool = Bool::True;
+//
+//     if bool == Bool::True {
+//         println!("foo");
+//     }
+// }
+// ```
+pub(crate) fn bool_to_enum(acc: &mut Assists, ctx: &AssistContext<'_>) -> Option<()> {
+    let BoolNodeData { target_node, name, ty_annotation, initializer, definition } =
+        find_bool_node(ctx)?;
+
+    let target = name.syntax().text_range();
+    acc.add(
+        AssistId("bool_to_enum", AssistKind::RefactorRewrite),
+        "Convert boolean to enum",
+        target,
+        |edit| {
+            if let Some(ty) = &ty_annotation {
+                cov_mark::hit!(replaces_ty_annotation);
+                edit.replace(ty.syntax().text_range(), "Bool");
+            }
+
+            if let Some(initializer) = initializer {
+                replace_bool_expr(edit, initializer);
+            }
+
+            let usages = definition
+                .usages(&ctx.sema)
+                .in_scope(&SearchScope::single_file(ctx.file_id()))
+                .all();
+            replace_usages(edit, &usages);
+
+            add_enum_def(edit, ctx, &usages, target_node);
+        },
+    )
+}
+
+struct BoolNodeData {
+    target_node: SyntaxNode,
+    name: ast::Name,
+    ty_annotation: Option<ast::Type>,
+    initializer: Option<ast::Expr>,
+    definition: Definition,
+}
+
+/// Attempts to find an appropriate node to apply the action to.
+fn find_bool_node(ctx: &AssistContext<'_>) -> Option<BoolNodeData> {
+    if let Some(let_stmt) = ctx.find_node_at_offset::<ast::LetStmt>() {
```

Comment:
We should look for `ctx.find_node_at_offset::<ast::Name>() ` first, and then check if that name is part of static, let, or const. Reason being we don't want the assist to be available when the cursor is in the initializer expressions of these constructs whi can be annoying. That is `let foo$0 = ...;` should have the assist be available, but not `let foo = bar::bar$0();`

Reply:
Ah okay, should be fixed now!